### PR TITLE
perf: core/TextReader for faster json ingestion

### PR DESCRIPTION
**Pull request by jasonk000**
### Description

Use a smaller default buffer in the TextReader during parsing. By default, the [LineIterator creates a BufferedReader if one is not provided](https://github.com/apache/commons-io/blob/master/src/main/java/org/apache/commons/io/LineIterator.java#L84-L88); the default [BufferedReader creates an 8KB buffer](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/9a751dc19fae78ce58fb0eb176522070c992fb6f/jdk/src/share/classes/java/io/BufferedReader.java#L88).

Improve this by creating a `FastLineIterator` which minimises the repeated creation of buffers.

Before vs after
```
Benchmark                         Mode  Cnt     Score    Error  Units
JsonLineReaderBenchmark.baseline  avgt   15  3459.871 ± 106.175  us/op      [before]
JsonLineReaderBenchmark.baseline  avgt   15  3022.053 ± 51.286  us/op       [after]
```
Existing tests cover the change (JsonLineReader, CSVReader). Replacement for #12302.

### Changes

- Introduces a new benchmark `JsonLineReaderBenchmark`
- Adds a `FastLineIterator` and switches the `TextReader` to use it.

This PR has:

- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jasonk000** [comment]
Middlemanager cluster performance shows significant boost after deployment
image

**Comment by jasonk000** [comment]
@samarthjain @maytasm could you please review or help me tag the correct reviewer for this?

File: core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.NoSuchElementException;
+
+/**
+ * Like the Apache Commons LineIterator, but faster.
+ */
+public class FastLineIterator implements CloseableIterator<String>
+{
+  static final int BUFFER_SIZE = 512;
```
Comment (samarthjain) [inline core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java:36]: All the static and instance variable fields should ideally be private unless there is a strong reason you want to relax their visibility.

File: core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    this.buffer.size(0);
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    nextLine = null;
+    finished = true;
+    source.close();
+    // Note: do not remove the thread local buffer; retain it for reuse later
```
Comment (samarthjain) [inline core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java:80]: Should the buffer be cleared out if threads are being re-used?

File: benchmarks/src/test/java/org/apache/druid/benchmark/JsonLineReaderBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                    new DimensionsSpec(DimensionsSpec.getDefaultSchemas(ImmutableList.of("bar", "foo"))),
+                    ColumnsFilter.all()
+            ),
+            source,
+            null
+    );
+  }
+
+  @Setup
+  public void prepareData() throws Exception
```
Comment (samarthjain) [inline benchmarks/src/test/java/org/apache/druid/benchmark/JsonLineReaderBenchmark.java:86]: Looking at the data here I can't tell if this is benchmarking cases where the String size is larger than the buffer.

File: core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java (hunk 1 above)
Comment (jasonk000) [reply core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java:36]: +1 for `CR`/`LF`.

For the buffer_size .. I use it in [tests such as this](https://github.com/apache/druid/pull/13545/files#diff-0a40f7c3d6223b3ada24ab09e70eb3f61c17728671de580dd96f22ea36d13ac5R171-R173), 

I can make it private but then we need to manually copy the value. Thoughts?

```
    // random lengths that force multiple buffer trips
    String line1 = RandomStringUtils.random(FastLineIterator.BUFFER_SIZE * 3 + 10);
    String line2 = RandomStringUtils.random(FastLineIterator.BUFFER_SIZE * 2 + 15);
    String line3 = RandomStringUtils.random(FastLineIterator.BUFFER_SIZE + 9);
```

File: core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java (hunk 2 above)
Comment (jasonk000) [reply core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java:80]: No need to, the buffer contents are overwritten when used.

File: benchmarks/src/test/java/org/apache/druid/benchmark/JsonLineReaderBenchmark.java (hunk 3 above)
Comment (jasonk000) [reply benchmarks/src/test/java/org/apache/druid/benchmark/JsonLineReaderBenchmark.java:86]: The provided data line is currently 753 lines long.

File: core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java (hunk 1 above)
Comment (samarthjain) [reply core/src/main/java/org/apache/druid/data/input/impl/FastLineIterator.java:36]: Sorry, I should have been more clear. I was referring to all the variables. 
```
static final byte CR = (byte) '\r';
static final byte LF = (byte) '\n';

final InputStream source;
final ByteArrayList buffer;

boolean finished;
String nextLine;
```

It is fine for constants like `BUFFER_SIZE` etc as non-private especially if you need to access them in tests.

**Review by maytasm — APPROVED** [review]
LGTM

**Comment by samarthjain** [comment]
Looks good, @jasonk000 . Just one comment - how about tests for cases where there is only CR or LF in a line?

**Pull request merged**